### PR TITLE
ADD: design doc for writing bypass TiDB

### PR DESCRIPTION
**Pull request by xuanyu66**
https://github.com/tidb-incubator/TiBigData/issues/163
- Checkpoint in this design has some data loss potential, so it is not supported now. We should optimize it in future

### CODE REVIEW

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+## Detailed Design
+
+### Add TiDBDynamicTableSink option
+
+In the factory method of TiDBDynamicTableFactory, we will add TiDBDynamicTableSink option. The user can choose how to write data by setting the Sink mode (JDBC or TiKV). 
+
+### Implement different dimensional transaction levels
+
+TiKV implements Google's Percolator transaction mode, which is a variation of 2pc. See [tidb-transaction-model](https://pingcap.com/zh/blog/tidb-transaction-model).
```
Comment (zhangyangyu) [inline docs/design/2022-03-10-write-bypass-tidb.md:50]: Maybe we could use an English description, like https://tikv.org/deep-dive/distributed-transaction/introduction/

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+This article introduces an optimized solution for writing Flink connector to TiDB cluster, by bypassing TiDB server and writing TiKV directly.
+
+## Motivation or Background
+
+Currently, Fink connector uses JdbcDynamicTableSink to write data through TiDB server.
+
+However, this approach leads to the following problems.
+- Lack of atomicity, if the Flink task fails and exists during the writing process, some data will be written successfully while the others not.
+- Lack of isolation, other transactions can see part of the written data during the writing process.
```
Comment (zhangyangyu) [inline docs/design/2022-03-10-write-bypass-tidb.md:31]: Gives me a feeling TiDB doesn't support transaction, which is not the fact.

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+It is worth noting that the box actually indicates the add of Transformation, and the operation of the Transformation is not executed immediately, but later executed in parallel in each TaskManager. Transformation includes Function and Operator. So the flowchart shows the order in which each Transformation is added.
+
+### MiniBatch mode
+
+ProcessElement is executed once when it iterates over each element. As you can see, MiniBatch has no global transaction guarantee and performs a 2pc write to the data whenever the number of data rows reaches a certain threshold.
+
+![image alt text](imgs/mini-batch.png)
+
+###Global
```
Comment (zhangyangyu) [inline docs/design/2022-03-10-write-bypass-tidb.md:82]:
```suggestion
### Global
```

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+### Checkpoint
+
+Unlike the previous two implementations, TIDBSinkFunction inherits TwoPhaseCommitSinkFunction, which provides excatly-once semantics. calls to the commit function occur at the checkpoint node, as described in [TwoPhaseCommitSinkFunction](https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.html).
+
+![image alt text](imgs/checkpoint.png)
+
+## Compatibility
+
+Flink Connector is forward compatible, the previous JDBC sink method still exists, and users can choose write-mode which they want.
```
Comment (zhangyangyu) [inline docs/design/2022-03-10-write-bypass-tidb.md:96]: I think it should be backward compatible.

File: docs/design/2022-03-10-write-bypass-tidb.md (hunk 2 above)
Comment (xuanyu66) [reply docs/design/2022-03-10-write-bypass-tidb.md:31]: Give a description that JdbcDynamicTableSink doesn't provide transaction semantics, not TiDB

**Comment by xuanyu66** [comment]
@zhangyangyu @humengyu2012 @shiyuhang0 PTAL

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+For each transaction commit, we select a row as the primary key and perform a two-stage commit, while the transaction satisfies the ACID principle.
+
+![image alt text](imgs/2pc-percolator.png)
+
+Transaction levels can be classified as Global, MiniBatch, Checkpoint:
+- Global is only applicable to bounded data streams, where all written data is considered as a whole and written to TiKV through one big transaction, which is not recommended if there are frequent write conflicts.
+- Checkpoint only supports data sources with checkpointing enabled, Flink uses each checkpoint as a transaction separator, and records commit transaction information through the checkpoint, which combined with TiKV's two-stage commit, can achieve the semantics of exactly once when writing TiDB upstream.
+- MiniBatch has no restriction on data sources, no global transaction semantics, and writes to TiDB with a single Task as the dimension. TiDB will be written when the data reaches a specified number of rows or a certain interval since the last write.
+
+### Data de-duplication
```
Comment (shiyuhang0) [inline docs/design/2022-03-10-write-bypass-tidb.md:61]: Do we need to introduce the behavior of conflicting with TiDB data

File: docs/design/2022-03-10-write-bypass-tidb.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+### Global
+
+Global prewrites the primary key when the stream is generated, then the ProcessElement only prewrites the Secondary Keys and commits the primary key once all the data has been written.
+
+![image alt text](imgs/global.png)
+
+### Checkpoint
+
+Unlike the previous two implementations, TIDBSinkFunction inherits TwoPhaseCommitSinkFunction, which provides exactly-once semantics. calls to the commit function occur at the checkpoint node, as described in [TwoPhaseCommitSinkFunction](https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.html).
```
Comment (shiyuhang0) [inline docs/design/2022-03-10-write-bypass-tidb.md:90]: whether we realize the exactly-once semantics. I don't see the rollback logic in the image

File: docs/design/2022-03-10-write-bypass-tidb.md (hunk 5 above)
Comment (xuanyu66) [reply docs/design/2022-03-10-write-bypass-tidb.md:61]: It's another topic about how data sink to the database, insert or upsert

File: docs/design/2022-03-10-write-bypass-tidb.md (hunk 6 above)
Comment (xuanyu66) [reply docs/design/2022-03-10-write-bypass-tidb.md:86]: It's abort method, which is not mentioned in the image

**Review by shiyuhang0 — APPROVED** [review]

**Review by humengyu2012 — APPROVED** [review]

**Pull request merged**